jedi/evaluate/compiled/subprocess/__init__.py

```diff
@@ -56,6 +56,23 @@ def _get_function(name):
     return getattr(functions, name)
 
 
+def _clean_process(process, thread):
+    try:
+        process.kill()
+        process.wait()
+    except (AttributeError, TypeError, OSError):
+        # If the Python process is terminating, it will remove some modules
+        # earlier than others and in general it's unclear how to deal with
+        # that so we just ignore the exceptions here.
```

Comment:
I like it a lot that you found `atexit`. I tried to search for exactly that, but couldn't find it. IMO this is way cleaner than what I'm doing here (what I'm doing here is just an aberration).

I feel like if we have that, there's no reason anymore to catch any errors. I'm only catching these errors, because `__del__` is called when some objects like the module `signal` have already been garbage collected. I feel like your approach is mostly right.

However if we want to get rid of catching the exceptions, we have to use something like

```
def __init__(self):
    self._cleanup_callback = functools.partial(_cleanup_process, process, t)
    atexit.register(self._cleanup_callback)

def _kill(self):
    self.is_crashed = True
    atexit.unregister(self._cleanup_callback)  # This is the important line
    self._cleanup_callback()
```

If we don't unregister, it will try to close everything again for a lot of potential processes - when Jed is running for a long time. I would like to avoid that. However this looks better IMO. Probably we also need to check if the process was already cleaned up.

If you don't want to do that - or have no time, feel free to leave it as it is. It's better than what we have before anyway :)

Reply:
> I feel like if we have that, there's no reason anymore to catch any errors.

I can try removing `AttributeError` and `TypeError` but I think `OSError` is still needed (I added it because I was seeing this exception at the end of tests on AppVeyor).

> If we don't unregister, it will try to close everything again for a lot of potential processes - when Jed is running for a long time.

The issue is that `atexit.unregister` is not available on Python 2. Let me figure something out.

Reply:
A solution is to use [`weakref.finalize`](https://docs.python.org/3.5/library/weakref.html#weakref.finalize) instead of `atexit`. It's not available on Python 2 so we need to backport it. The backport code is from https://gist.github.com/florentx/52dc3d00feb513d10838 (which is based on the official implementation of `weakref.finalize`). PR is updated.

Reply:
Ok. How about only using atexit.unregister for Python3? I feel like it's a clean way and Python 2 is leaving this world in a year :)

Reply:
This would obviously mean that we still have to have the solution we're currently using for Python2, but that's fine IMO.